### PR TITLE
Fix :meth:`BarChart.change_bar_values`  not updating when height is 0.

### LINKED ISSUE #2636: `BarChart` bars of height 0 do not get updated as expected.
## Description of bug / unexpected behavior

`BarChart` bars of height do not get updated as expected.

## Expected behavior
I expected to be able to updated `BarChart` bars using the provided `change_bar_values` method, but this appears not to work for bars of height zero.

## How to reproduce the issue

In the following, the bars are not updated in the `Foo` scene (initial height 0.0), whereas they are updated in the `Bar` scene (initial height 0.0001).

<details><summary>Code for reproducing the problem</summary>

```py
class Foo(Scene):
    def construct(self):
        values=[0 for _ in range(11)]

        chart = BarChart(
            values,
            y_range=[0, 10, 1],
            y_axis_config={"font_size": 24},
        )
        new_values=[1 for _ in range(11)]
        chart.change_bar_values(new_values)
        self.add(chart)


class Bar(Scene):
    def construct(self):
        values=[0.0001 for _ in range(11)]

        chart = BarChart(
            values,
            y_range=[0, 10, 1],
            y_axis_config={"font_size": 24},
        )
        new_values=[1 for _ in range(11)]
        chart.change_bar_values(new_values)
        self.add(chart)
```

</details>

## Additional media files

<details><summary>Images/GIFs</summary>

Foo:

Foo_ManimCE_v0 15 1

Bar:

Bar_ManimCE_v0 15 1

</details>

## System specifications

<details><summary>System Details</summary>

```
Manjaro Linux 21.2.5
Python 3.10.2
manim 0.15.1
```

</details>

### PR DESCRIPTION
**Pull request by hydrobeam**
## Overview: What does this pull request change?

Closes #2636 

- Made it so that if the previous value of a bar chart is zero during `change_bar_values` is zero (caught during  a try/except), then a new bar is created and replaces its position in the `self.bars` `VGroup`
- Added `self.insert` methods to `Mobject` and `OpenGLMobject` to accomodate this.
- Refactored out bar creation into `BarChart._create_bar` for use in `BarChart._add_bars`  and `change_bar_values`
- Added `self._update_colors` since if new bars are created, then their colour won't match the rest of the bar chart, so the colours will be re-initalized.

```py
class Foo(Scene):
    def construct(self):
        values=[0 for _ in range(11)]

        chart = BarChart(
            values,
            y_range=[0, 10, 1],
            y_axis_config={"font_size": 24},
        )
        new_values=[1 for _ in range(11)]
        chart.change_bar_values(new_values)
        self.add(chart)
```

Before:
Foo_ManimCE_v0 14 0

After:
Foo_ManimCE_v0 14 0

## Further Information and Comments

Currently missing proper tests. Also, there might be value in having a singular `change_bar_value`.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/mobject/graphing/probability.py [hunk 1]
```diff
@@ -304,6 +304,19 @@ def __init__(
...
+
+        Sets the color of ``self.bars`` via ``self.bar_colors``.
+
+        Primarily used when the bars are initialized with ``self._add_bars``
+        or updated via ``self.change_bar_values``.
+        """
+        if isinstance(self.bar_colors, str):
+            self.bars.set_color_by_gradient(self.bar_colors)
+        else:
+            self.bars.set_color_by_gradient(*self.bar_colors)
```
Comment (Kiran-Raj-Dev) [inline manim/mobject/graphing/probability.py:318]: Given that the type hint for `bar_colors` also supports `str` I think this is fine for that but since `bar_colors` is plural can we not just tell the user to always enter the values in a list? 
And remove this function entirely and just put in `self.bars.set_color_by_gradient(*self.bar_colors)` in place of where it is called?

**Review by Kiran-Raj-Dev — DISMISSED** [review]
Cool! Thanks for the fix!
Left a suggestion and a potential design change and whether you think that would be a good idea.

File: manim/mobject/graphing/probability.py [hunk 2]
```diff
@@ -283,9 +283,9 @@ def __init__(
             (x_axis_config,), (kwargs.pop("x_axis_config", None),)
         )
 
-        self.bars = None
-        self.x_labels = None
-        self.bar_labels = None
+        self.bars: VGroup
+        self.x_labels: VGroup
```
Comment (MrDiver) [inline manim/mobject/graphing/probability.py:287]: I find this change rather suspicious. Because it removes the default value for the attribute so that it basically doesn't exist. I would change this to `self.x: Optional[VGroup] = None` for the three attributes because else it changes functionality.

File: manim/mobject/graphing/probability.py [hunk 3]
```diff
@@ -339,27 +352,48 @@ def _add_x_axis_labels(self):
...
+        Rectangle
+            A positioned rectangle representing a bar on the chart.
+        """
+
+        # bar measurements relative to the axis
+
+        # distance from between the y-axis and the top of the bar
+        bar_h = abs(self.c2p(0, value)[1] - self.c2p(0, 0)[1])
+        # width of the bar
+        bar_w = self.c2p(self.bar_width, 0)[0] - self.c2p(0, 0)[0]
```
[the remark this answers is not in the document]
Comment (MrDiver) [reply manim/mobject/graphing/probability.py:376]: Looks good to me!

File: manim/mobject/graphing/probability.py [hunk 4]
```diff
@@ -339,27 +352,48 @@ def _add_x_axis_labels(self):
...
-            )
-
-            pos = UP if (value >= 0) else DOWN
-            bar.next_to(self.c2p(i + 0.5, 0), pos, buff=0)
-            self.bars.add(bar)
-        if isinstance(self.bar_colors, str):
-            self.bars.set_color_by_gradient(self.bar_colors)
-        else:
-            self.bars.set_color_by_gradient(*self.bar_colors)
+            temp_bar = self._create_bar(bar_number=i, value=value)
```
Comment (MrDiver) [inline manim/mobject/graphing/probability.py:393]: This is more of a personal taste but i think `tmp` is the short for `temporary`. `temp` sounds more like `temperature` at least that is what i saw most often in production code.

File: manim/mobject/graphing/probability.py [hunk 5]
```diff
@@ -339,27 +352,48 @@ def _add_x_axis_labels(self):
...
+            self.bars.add(temp_bar)
 
+        self._update_colors()
         self.add_to_back(self.bars)
 
     def get_bar_labels(
```
Comment (MrDiver) [inline manim/mobject/graphing/probability.py:399]: Well, it isn't really changed by this PR but the typing for this constructor is wrong for `label_constructor` it should be `Type[VMobject]` instead of `VMobject` see https://docs.python.org/3/library/typing.html#typing.Type

File: manim/mobject/graphing/probability.py (hunk 2 above)
Comment (MrDiver) [reply manim/mobject/graphing/probability.py:287]: Except for self.bars which might be better of just being `self.bars: VGroup = VGroup()` and then remove the initialization line in `_add_bars` because it's only called in `__init__` anyway

File: manim/mobject/graphing/probability.py [hunk 6]
```diff
@@ -468,11 +504,17 @@ def construct(self):
                     # if already negative, then we move the bottom edge of the bar
                     # to the location of the previous top
 
-                bar.stretch_to_fit_height(quotient * bar.height)
+                bar.stretch_to_fit_height(abs(quotient) * bar.height)
```
Comment (MrDiver) [inline manim/mobject/graphing/probability.py:507]: What exactly does this fix? : `abs(quotient)`

Comment (MrDiver) [reply manim/mobject/graphing/probability.py:507]: Well i guess it fixes bars with negative values

File: manim/mobject/graphing/probability.py [hunk 7]
```diff
@@ -468,11 +504,17 @@ def construct(self):
...
 
             except ZeroDivisionError:
```
Comment (MrDiver) [inline manim/mobject/graphing/probability.py:509]: Can we change this to not use try/except. I mean we know when theres a division by zero error. No need to pull up a whole try/except. if else should do the job 🤷🏼

File: manim/mobject/mobject.py [hunk 8]
```diff
@@ -425,13 +425,34 @@ def add(self, *mobjects: "Mobject"):
...
+        self.submobjects.insert(index, mobject)
+
     def __add__(self, mobject):
         raise NotImplementedError
 
     def __iadd__(self, mobject):
         raise NotImplementedError
 
-    def add_to_back(self, *mobjects: "Mobject"):
+    def add_to_back(self, *mobjects: Mobject):
```
Comment (MrDiver) [inline manim/mobject/mobject.py:455]: I mean im all for typehint changes. But they don't really seem to be related to this PR 🤔 can we just move them to a different PR instead?
@behackl We should also setup a typeguide in general or discuss this in the team which kind of typing we want to use. Because else it's just a mess.

**Review by MrDiver — CHANGES_REQUESTED** [review]

File: manim/mobject/mobject.py (hunk 8 above)
Comment (MrDiver) [reply manim/mobject/mobject.py:455]: - #2765

**Comment by MrDiver** [comment]
```py
from manim import *

class Foo(Scene):
    def construct(self):
        values=[0 for _ in range(11)]+[1,1,1]

        chart = BarChart(
            values,
            y_range=[0, 10, 1],
            y_axis_config={"font_size": 24},
        )
        new_values=[i for i in range(11)] + [-2,-3,-4]
        self.add(chart)
        self.play(chart.animate.change_bar_values(new_values))

```

Just a different flavor of test.

File: manim/mobject/graphing/probability.py (hunk 7 above)
Comment (hydrobeam) [reply manim/mobject/graphing/probability.py]: Sure, done.

File: manim/mobject/graphing/probability.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hydrobeam) [reply manim/mobject/graphing/probability.py:381]: Good suggestion, done.

File: manim/mobject/mobject.py (hunk 8 above)
Comment (hydrobeam) [reply manim/mobject/mobject.py:455]: It's been a while since this PR, but I think most of these changes are because of adding `from __future__ import annotations` at the top, which removes the need for quotes around certain types and allows the use of builtins. There's a good chance they were automated 🤔.

They aren't really relevant but I don't think they hurt to have. I'd be willing to move these to another PR for the sake of a cleaner history though.

File: manim/mobject/graphing/probability.py (hunk 1 above)
Comment (hydrobeam) [reply manim/mobject/graphing/probability.py]: I like the suggestion of forcing `bar_colors` to be inputted as a list. but I think having colour updating in a separate function is clearer about conveying its purpose, even if it looks a little silly in its current state 🤷

**Review by hydrobeam — COMMENTED** [review]
Implemented most of the suggested changes, with some minor type/doc adjustments and a deprecation. 

I also deprecated passing in a string to `bar_colors` (unfortunately rather manually, since our `deprecated` module doesn't support this type of deprecation {tried quite a bit to make `deprecated_params` work})

**Comment by MrDiver** [comment]
Please review requested changes!

**Review by MrDiver — APPROVED** [review]
